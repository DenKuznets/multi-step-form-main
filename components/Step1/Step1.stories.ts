import type { Meta, StoryObj } from '@storybook/react';
import  Step1  from './Step1';

const meta = {
  title: 'Step1',
  component: Step1,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  parameters: {
    // More on how to position stories at: https://storybook.js.org/docs/react/configure/story-layout
    layout: 'fullscreen',
  },
} satisfies Meta<typeof Step1>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
