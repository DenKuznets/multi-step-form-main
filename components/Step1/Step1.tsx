'use client';
import { useForm, SubmitHandler } from 'react-hook-form';
import { useAppDispatch, useAppSelector } from '@/lib/redux/hooks';
import { selectName } from '@/lib/redux/slices/appSlice';
import { selectPhone } from '@/lib/redux/slices/appSlice';
import { selectEmail } from '@/lib/redux/slices/appSlice';
import { useRouter } from 'next/navigation';
import { setName, setEmail, setPhone } from '@/lib/redux/slices/appSlice';

interface FormValues {
    name: string;
    phone: string;
    email: string;
}

const Step1 = () => {
    const router = useRouter();
    const name = useAppSelector(selectName);
    const phone = useAppSelector(selectPhone);
    const email = useAppSelector(selectEmail);
    const dispatch = useAppDispatch();
    const { formState, register, handleSubmit } = useForm<FormValues>({
        defaultValues: {
            name: name,
            email: email,
            phone: phone
        }
    });
    const { errors } = formState;
    const onSubmit: SubmitHandler<FormValues> = (data) => {
        console.log('submit', data);
        dispatch(setName(data.name));
        dispatch(setPhone(data.phone));
        dispatch(setEmail(data.email));
        router.push('/select-plan');
    };

    return (
        <form
            noValidate
            className="flex w-full flex-col gap-2 sm:gap-4"
            aria-label="personal info form"
            onSubmit={handleSubmit(onSubmit)}
        >
            <div className="relative">
                <p className="absolute right-0 text-[0.75rem] font-bold text-strawberryRed sm:text-[0.9rem]">
                    {errors.name?.message}
                </p>
                <label htmlFor="name" className={`label`}>
                    Name
                </label>
                <input
                    id="name"
                    className={`input  ${
                        errors.name ? 'outline-strawberryRed' : ''
                    }`}
                    placeholder="e.g. Stephen King"
                    {...register(`name`, {
                        required: 'This field is required'
                    })}
                />
            </div>
            <div className="relative">
                <p className="absolute right-0 text-[0.75rem] font-bold text-strawberryRed sm:text-[0.9rem]">
                    {errors.email?.message}
                </p>
                <label htmlFor="email" className="label">
                    Email Address
                </label>
                <input
                    id="email"
                    placeholder="e.g. stephenking@lorem.com"
                    className={`input  ${
                        errors.email ? 'outline-strawberryRed' : ''
                    }`}
                    {...register(`email`, {
                        required: 'This field is required',
                        pattern: {
                            value: /^[\w-.]+@([\w-]+\.)+[\w-]{2,4}$/g,
                            message: 'Invalid email'
                        }
                    })}
                />
            </div>
            <div className="relative">
                <p className="absolute right-0 text-[0.75rem] font-bold text-strawberryRed sm:text-[0.9rem]">
                    {errors.phone?.message}
                </p>
                <label htmlFor="phone" className="label">
                    Phone Number
                </label>
                <input
                    id="phone"
                    className={`input  ${
                        errors.phone ? 'outline-strawberryRed' : ''
                    }`}
                    placeholder="e.g. +1 234 567 890"
                    {...register(`phone`, {
                        required: 'This field is required'
                    })}
                />
            </div>

            <div className="fixed bottom-0 left-0 flex w-full justify-end bg-white p-4 sm:absolute sm:mb-4 sm:p-0">
                <button className="btn btn-next ">next step</button>
            </div>
        </form>
    );
};

export default Step1;
