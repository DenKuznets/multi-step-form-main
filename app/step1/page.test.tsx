import '@testing-library/jest-dom';
import Step1 from './page';
import {
    renderWithProviders,
    screen,
    within
} from '@/lib/redux/utils-for-tests';

test('renders correctly', () => {
    renderWithProviders(<Step1 />);

    const h1 = screen.getByRole('heading', { level: 1 });
    const info = screen.getByText(
        'Please provide your name, email address, and phone number.'
    );
    const form = screen.getByRole('form', { name: /personal info form/i });
    const nameLabel = within(form).getByText(/name/i);
    const emailLabel = within(form).getByText(/email address/i);
    const phoneLabel = within(form).getByText(/phone number/i);
    const nameInput = screen.getByRole('textbox', { name: /name/i });
    const emailInput = screen.getByRole('textbox', { name: /email address/i });
    const phoneInput = screen.getByRole('textbox', { name: /phone number/i });
    const nextStepButton = screen.getByRole('button', {
        name: /next step/i
    });
    
    expect(h1).toBeInTheDocument();
    expect(info).toBeInTheDocument();
    expect(form).toBeInTheDocument();
    expect(nameLabel).toBeInTheDocument();
    expect(emailLabel).toBeInTheDocument();
    expect(phoneLabel).toBeInTheDocument();
    expect(nameInput).toBeInTheDocument();
    expect(emailInput).toBeInTheDocument();
    expect(phoneInput).toBeInTheDocument();
    expect(nextStepButton).toBeInTheDocument();
});

test('renders Home unchanged', () => {
    const { container } = renderWithProviders(<Step1 />);
    expect(container).toMatchSnapshot();
});
