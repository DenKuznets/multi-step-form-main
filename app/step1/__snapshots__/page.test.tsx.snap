// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Home unchanged 1`] = `
<div>
  <div
    class="step sm:relative sm:h-full"
  >
    <h1
      class="step-header"
    >
      Personal info
    </h1>
    <p
      class="step-info"
    >
      Please provide your name, email address, and phone number.
    </p>
    <form
      aria-label="personal info form"
      class="flex w-full flex-col gap-2 sm:gap-4 "
      novalidate=""
    >
      <div
        class="relative"
      >
        <p
          class="text-strawberryRed absolute   right-0 text-[0.75rem] font-bold sm:text-[0.9rem]"
        />
        <label
          class="label"
          for="name"
        >
          Name
        </label>
        <input
          class="input  "
          id="name"
          name="name"
          placeholder="e.g. stephenking@lorem.com"
        />
      </div>
      <div
        class="relative"
      >
        <p
          class="text-strawberryRed absolute   right-0 text-[0.75rem] font-bold sm:text-[0.9rem]"
        />
        <label
          class="label"
          for="email"
        >
          Email Address
        </label>
        <input
          class="input  "
          id="email"
          name="email"
          placeholder="e.g. Stephen King"
        />
      </div>
      <div
        class="relative"
      >
        <p
          class="text-strawberryRed absolute   right-0 text-[0.75rem] font-bold sm:text-[0.9rem]"
        />
        <label
          class="label"
          for="phone"
        >
          Phone Number
        </label>
        <input
          class="input  "
          id="phone"
          name="phone"
          placeholder="e.g. +1 234 567 890"
        />
      </div>
      <div
        class="fixed bottom-0 left-0 flex h-[4.5rem] w-full justify-end bg-white p-4 sm:absolute sm:min-h-[6rem] sm:items-center sm:pb-0 sm:pt-6 md:pr-[5.5rem]"
      >
        <button
          class="btn btn-next "
        >
          next step
        </button>
      </div>
    </form>
  </div>
</div>
`;
